From a3997e3dabd78a96d320b80695c458a0bbdf798f Mon Sep 17 00:00:00 2001
From: Jiamei Xie <jiamei.xie@arm.com>
Date: Wed, 19 Jul 2023 13:54:24 +0800
Subject: [PATCH] core:arm:add rdkronos platform

These changes are to add rdkronos platform to OP-TEE core.

Upstream-Status: Pending [Not submitted to upstream yet]
Signed-off-by: Jun Zhang <jun.zhang@arm.com>
Signed-off-by: Jiamei Xie <jiamei.xie@arm.com>
Signed-off-by: Ziad Elhanafy <ziad.elhanafy@arm.com>

%% original patch: 0002-core-arm-add-rdkronos-platform.patch
---
 core/arch/arm/plat-rdkronos/conf.mk           | 30 ++++++++++++
 core/arch/arm/plat-rdkronos/main.c            | 39 +++++++++++++++
 core/arch/arm/plat-rdkronos/platform_config.h | 47 +++++++++++++++++++
 .../arm/plat-rdkronos/rdkronos_core_pos.S     | 32 +++++++++++++
 core/arch/arm/plat-rdkronos/sub.mk            |  3 ++
 5 files changed, 151 insertions(+)
 create mode 100644 core/arch/arm/plat-rdkronos/conf.mk
 create mode 100644 core/arch/arm/plat-rdkronos/main.c
 create mode 100644 core/arch/arm/plat-rdkronos/platform_config.h
 create mode 100644 core/arch/arm/plat-rdkronos/rdkronos_core_pos.S
 create mode 100644 core/arch/arm/plat-rdkronos/sub.mk

diff --git a/core/arch/arm/plat-rdkronos/conf.mk b/core/arch/arm/plat-rdkronos/conf.mk
new file mode 100644
index 000000000..ae81c9f4c
--- /dev/null
+++ b/core/arch/arm/plat-rdkronos/conf.mk
@@ -0,0 +1,30 @@
+include core/arch/arm/cpu/cortex-armv8-0.mk
+
+# ARM debugger needs this
+platform-cflags-debug-info = -gdwarf-4
+platform-aflags-debug-info = -gdwarf-4
+
+CFG_DEBUG_INFO = y
+CFG_TEE_CORE_LOG_LEVEL = 4
+
+CFG_ARM64_core ?= y
+
+CFG_ARM_GICV3 = y
+
+CFG_CORE_SEL1_SPMC = y
+CFG_WITH_ARM_TRUSTED_FW	= y
+CFG_CORE_RESERVED_SHM = n
+
+$(call force,CFG_GIC,y)
+$(call force,CFG_PL011,y)
+$(call force,CFG_SECURE_TIME_SOURCE_CNTPCT,y)
+$(call force,CFG_CORE_ARM64_PA_BITS,42)
+$(call force,CFG_TEE_CORE_NB_CORE,16)
+
+CFG_CORE_HEAP_SIZE = 0x32000
+
+CFG_TEE_CORE_NB_CORE ?= 16
+
+CFG_TZDRAM_START ?= 0xfe000000
+CFG_TZDRAM_SIZE  ?= 0x02000000
+
diff --git a/core/arch/arm/plat-rdkronos/main.c b/core/arch/arm/plat-rdkronos/main.c
new file mode 100644
index 000000000..f3f0ca540
--- /dev/null
+++ b/core/arch/arm/plat-rdkronos/main.c
@@ -0,0 +1,39 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2023, Arm Limited
+ */
+
+#include <console.h>
+#include <drivers/gic.h>
+#include <drivers/pl011.h>
+#include <kernel/boot.h>
+#include <mm/core_mmu.h>
+#include <platform_config.h>
+#include <stdint.h>
+#include <trace.h>
+
+static struct pl011_data console_data __nex_bss;
+
+register_ddr(DRAM0_BASE, DRAM0_SIZE);
+register_ddr(DRAM1_BASE, DRAM1_SIZE);
+
+register_phys_mem_pgdir(MEM_AREA_IO_SEC, CONSOLE_UART_BASE, PL011_REG_SIZE);
+register_phys_mem_pgdir(MEM_AREA_IO_SEC, GICD_BASE, GIC_DIST_REG_SIZE);
+register_phys_mem_pgdir(MEM_AREA_IO_SEC, GICC_BASE, GIC_CPU_REG_SIZE);
+
+void main_init_gic(void)
+{
+	gic_init(GICC_BASE, GICD_BASE);
+}
+
+void main_secondary_init_gic(void)
+{
+	gic_cpu_init();
+}
+
+void console_init(void)
+{
+	pl011_init(&console_data, CONSOLE_UART_BASE, CONSOLE_UART_CLK_IN_HZ,
+		   CONSOLE_BAUDRATE);
+	register_serial_console(&console_data.chip);
+}
diff --git a/core/arch/arm/plat-rdkronos/platform_config.h b/core/arch/arm/plat-rdkronos/platform_config.h
new file mode 100644
index 000000000..e398ccc5b
--- /dev/null
+++ b/core/arch/arm/plat-rdkronos/platform_config.h
@@ -0,0 +1,47 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2023, Arm Limited
+ */
+
+#ifndef PLATFORM_CONFIG_H
+#define PLATFORM_CONFIG_H
+
+#include <mm/generic_ram_layout.h>
+
+/* Make stacks aligned to data cache line length */
+#define STACK_ALIGNMENT		64
+
+#define UART0_BASE		0x2a400000
+#define UART1_BASE		0x2a410000
+#define CONSOLE_UART_BASE	UART1_BASE
+
+#define DRAM0_BASE		0x80000000
+#define DRAM0_SIZE		0x80000000
+
+#define DRAM1_BASE		0x8080000000ULL
+#define DRAM1_SIZE		0x80000000ULL
+
+/* GIC related constants */
+#define GICD_BASE		UL(0x30000000)
+#define GICR_BASE		UL(0x301C0000)
+#define GICC_BASE		UL(0x2C000000)
+
+#define UART_BAUDRATE		115200
+#define CONSOLE_BAUDRATE	UART_BAUDRATE
+#define CONSOLE_UART_CLK_IN_HZ	UL(7372800)
+
+#define SYS_COUNTER_FREQ_IN_TICKS	UL(7372800)
+
+ /* KRONOS topology related constants */
+#define KRONOS_MAX_CPUS_PER_CLUSTER		U(1)
+#define PLAT_ARM_CLUSTER_COUNT			U(1)
+#define PLAT_KRONOS_CHIP_COUNT			U(1)
+#define KRONOS_MAX_CLUSTERS_PER_CHIP	U(16)
+#define KRONOS_MAX_PE_PER_CPU			U(1)
+
+#define PLATFORM_CORE_COUNT	(PLAT_KRONOS_CHIP_COUNT * \
+							PLAT_ARM_CLUSTER_COUNT * \
+							KRONOS_MAX_CPUS_PER_CLUSTER * \
+							KRONOS_MAX_PE_PER_CPU)
+
+#endif /*PLATFORM_CONFIG_H*/
diff --git a/core/arch/arm/plat-rdkronos/rdkronos_core_pos.S b/core/arch/arm/plat-rdkronos/rdkronos_core_pos.S
new file mode 100644
index 000000000..127135bd3
--- /dev/null
+++ b/core/arch/arm/plat-rdkronos/rdkronos_core_pos.S
@@ -0,0 +1,32 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2023, Arm Limited
+ */
+
+#include <asm.S>
+#include <arm.h>
+#include "platform_config.h"
+
+FUNC get_core_pos_mpidr , :
+        mov     x4, x0
+
+        /*
+         * The MT bit in MPIDR is always set for n1sdp and the
+         * affinity level 0 corresponds to thread affinity level.
+         */
+
+        /* Extract individual affinity fields from MPIDR */
+        ubfx    x0, x4, #MPIDR_AFF0_SHIFT, #MPIDR_AFFINITY_BITS
+        ubfx    x1, x4, #MPIDR_AFF1_SHIFT, #MPIDR_AFFINITY_BITS
+        ubfx    x2, x4, #MPIDR_AFF2_SHIFT, #MPIDR_AFFINITY_BITS
+        ubfx    x3, x4, #MPIDR_AFF3_SHIFT, #MPIDR_AFFINITY_BITS
+
+        /* Compute linear position */
+        mov     x4, #KRONOS_MAX_CLUSTERS_PER_CHIP
+        madd    x2, x3, x4, x2
+        mov     x4, #KRONOS_MAX_CPUS_PER_CLUSTER
+        madd    x1, x2, x4, x1
+        mov     x4, #KRONOS_MAX_PE_PER_CPU
+        madd    x0, x1, x4, x0
+        ret
+END_FUNC get_core_pos_mpidr
\ No newline at end of file
diff --git a/core/arch/arm/plat-rdkronos/sub.mk b/core/arch/arm/plat-rdkronos/sub.mk
new file mode 100644
index 000000000..63c999ed9
--- /dev/null
+++ b/core/arch/arm/plat-rdkronos/sub.mk
@@ -0,0 +1,3 @@
+global-incdirs-y += .
+srcs-y += main.c
+srcs-y	+= rdkronos_core_pos.S
-- 
2.25.1
